src/memory.jl

```diff
@@ -8,22 +12,21 @@ using ..CUDAdrv
 import ..@apicall, ..CuStream_t, ..CuDevice_t
 
 
-## buffer type
+## abstract buffer type
 
-struct Buffer
-    ptr::CuPtr{Cvoid}
-    bytesize::Int
+abstract type Buffer <: Ref{Cvoid} end
 
-    ctx::CuContext
-end
+# expected interface:
+# - similar()
+# - ptr, bytesize and ctx fields
+# - unsafe_convert() to certain pointers
 
-Base.unsafe_convert(::Type{<:Ptr}, buf::Buffer) =
-    throw(ArgumentError("cannot take the CPU address of a GPU buffer"))
-Base.unsafe_convert(::Type{CuPtr{T}}, buf::Buffer) where {T} = convert(CuPtr{T}, buf.ptr)
+Base.pointer(buf::Buffer) = buf.ptr
+Base.sizeof(buf::Buffer) = buf.bytesize
 
-function view(buf::Buffer, bytes::Int)
-    bytes > buf.bytesize && throw(BoundsError(buf, bytes))
-    return Mem.Buffer(buf.ptr+bytes, buf.bytesize-bytes, buf.ctx)
+function Base.view(buf::Buffer, bytes::Int)
```

Comment:
So if I understand correctly, at this lower-level, the method to slice into page-locked buffers would be something like:

```
# d_buff : device buffer
# h_buff : host buffer
# nstrides : number of slices
# stridebytes : size of slice in bytes
unsafe_copyto!(d_buff, view(h_buff, nstrides*stridebytes), stridebytes,
             stream=X) 
```